Search: Add a Gutenberg powered customizer

jsnmoon opened the pull request:
Follow-up for #20282, spun off from #20209. NOTE: The Customberg experience requires WordPress 5.8!

#### Changes proposed in this Pull Request:
<img width="1792" alt="Screen Shot 2021-06-30 at 5 12 13 PM" src="https://user-images.githubusercontent.com/4044428/124042554-518ed800-d9c6-11eb-80c3-10cb06aebffd.png">

Adds a React application that enables Gutenberg-styled customization of the Jetpack Search interface.

#### Jetpack product discussion
See pcNPJE-7J-p2.

#### Does this pull request change what data or activity we track or use?
Yes, we add additional Tracks analytics for the initial page view and when the save button is clicked.

#### Testing instructions:
* Apply this change to your site with a Jetpack Search subscription. 
* Navigate to the new customization page at `/wp-admin/admin.php?page=jetpack-search-customize`.

**On WordPress versions greater than 5.8**
* Ensure that the page renders a Gutenberg layout with a header, a sidebar to the right, and the Jetpack Search application in the center/left.
* Ensure that you can customize the search application via the controls in the options tab within the sidebar.
* Ensure that the "Jetpack Search" tab renders a description of customization interface and includes a link to the WordPress Customizer.
* Try resizing the page. Ensure that the page works as expected in all viewports ranging from mobile to desktop.

**On WordPress versions less than 5.8, including 5.8 RC's**
* Ensure that you're redirected to the Jetpack Search section of the Customizer.

Commit:
Add and use @wordpress/babel-preset-default

Commit:
Add Customberg application

Commit:
Changelog

Commit:
Add custom lint rules for Customberg, fix lint errors

bluefuton commented:
Great stuff @jsnmoon! Code looks good on first pass.

When I build the branch and navigate to `/wp-admin/admin.php?page=jetpack-search-customize`, I see:

<img width="388" alt="Screen Shot 2021-07-15 at 14 22 48" src="https://user-images.githubusercontent.com/17325/125718083-b196bdfe-eddd-4af4-beae-8c946a19d85e.png">

This is using `pnpm build && pnpm watch-search`. Let me know if there are other build steps I've missed.

kangzj commented on projects/plugins/jetpack/modules/search/customberg/components/header/index.jsx:
```diff
@@ -0,0 +1,47 @@
+	const isMediumViewport = useViewportMatch( 'medium' );
+
+	return (
+		<>
```
Is the Fragment necessary here?

jsnmoon replied:
Nope, good catch!

kangzj commented on projects/plugins/jetpack/modules/search/customberg/components/sidebar/color-control.jsx:
```diff
@@ -0,0 +1,42 @@
+export default function ColorControl( { disabled, value, onChange } ) {
+	return (
+		<div className="jp-search-customize-color-input components-base-control">
+			<div className="jp-search-customize-color-input-label">
```
`jp-search-customize-color-input-label` => `jp-search-customize-color-input__label`? not sure.

kangzj commented on projects/plugins/jetpack/modules/search/customberg/components/sidebar/excluded-post-types-control.jsx:
```diff
@@ -0,0 +1,59 @@
+
+/* eslint-disable react/jsx-no-bind */
+
+const VALID_POST_TYPES = global.JetpackInstantSearchOptions.postTypes;
```
Could we use it as params or in state store or anything, using globals directly is pretty unfriendly for unit testing.

jsnmoon replied:
Converted to a component property with the global value as a default, which should enable straightforward unit testing.

kangzj commented on projects/plugins/jetpack/modules/search/customberg/components/sidebar/index.jsx:
```diff
@@ -0,0 +1,113 @@
+ * Sidebar tabs.
+ *
+ * @param {object} props - Component properties
+ * @param {string} props.currentArea - Current area.
```
Do we fancy `currentTab` maybe?

jsnmoon replied:
👍 I prefer `currentTab`

kangzj commented on projects/plugins/jetpack/modules/search/customberg/components/sidebar/sidebar-options.jsx:
```diff
@@ -0,0 +1,134 @@
+	return (
+		<Panel
+			className={ classNames( 'jp-search-customize-sidebar-options', {
+				'jp-search-customize-sidebar-options--is-disabled': isDisabled,
```
`BEM` or Calypso standard is a question here. As @bluefuton suggested, we should follow [this](https://github.com/Automattic/wp-calypso/blob/trunk/docs/coding-guidelines/css.md), which is similar but different to `BEM`. Modifiers would be just like `is-disabled` if I haven't got it wrong.

jsnmoon replied:
Let's go with BEM as much as possible. I can imagine modifier classes like `is-disabled` wreaking havoc via style conflicts in the future.

jsnmoon commented:
> When I build the branch and navigate to `/wp-admin/admin.php?page=jetpack-search-customize`, I see:
> 
> <img alt="Screen Shot 2021-07-15 at 14 22 48" width="388" src="https://user-images.githubusercontent.com/17325/125718083-b196bdfe-eddd-4af4-beae-8c946a19d85e.png">
> 
> This is using `pnpm build && pnpm watch-search`. Let me know if there are other build steps I've missed. (@bluefuton)

Interesting. Just `pnpm watch-search` should be enough to build the assets, so this looks like a bug with Customberg. Any idea what the object it's trying to execute as a function?

Commit:
Remove unnecessary fragment

Commit:
Use valid post types as a component property

Commit:
Use currentTab instead of currentArea

Commit:
Enforce BEM more strictly

Commit:
Remove ES5 validation for search scripts

Commit:
Remove ES5 validation entirely :)

This should be temporary

kangzj commented:
> Interesting. Just `pnpm watch-search` should be enough to build the assets, so this looks like a bug with Customberg. Any idea what object it's trying to execute as a function? (I'm unfortunately not able to repro.)

Hi @jsnmoon  I tried to remove all node_modules and re-install but this happens for me as well still. The error is throwing from this line, 

![image](https://user-images.githubusercontent.com/1425433/125918024-0edd4214-c390-41fa-bf83-dc75a1b13f1d.png)

Commit:
Remove unused code and dependency

Commit:
Fix typo

Commit:
Bump dependency versions (again)

Commit:
Remove dependency extraction plugin

Commit:
Restore dependency extraction plugin

Commit:
Add WP 5.8 check

Commit:
Print redirect in the footer

Commit:
Merge branch 'master' of github.com:Automattic/jetpack into add/search-customberg-2

Commit:
Restore ES5 validation

This change is outside the scope of the PR, so we'll avoid doing this here.

Commit:
Merge branch 'master' of github.com:Automattic/jetpack into add/search-customberg-2

Commit:
Add code comment

Commit:
Don't parallelize search build, will fail ES5 test

Commit:
Merge branch 'master' of github.com:Automattic/jetpack into add/search-customberg-2

Commit:
Change requirement to WP 5.8

Commit:
Revert "Bump dependency versions (again)"

This reverts commit 932a515a6c806b1e48c30efe0963909eb5d17f31.

kangzj commented on projects/plugins/jetpack/modules/search/customberg/hooks/use-search-options.js:
```diff
@@ -0,0 +1,73 @@
+		'jetpack_search_excluded_post_types'
+	);
+	// Excluded Post Types is stored as a CSV string in site options. Convert into
+	const excludedPostTypes = useMemo( () => excludedPostTypesCsv?.split( ',' ), [
```
As the initial value for `jetpack_search_excluded_post_types` could be an empty string, when splitting on which, it produces `[""]` and breaks the size limit check.
![image](https://user-images.githubusercontent.com/1425433/127250687-8e7682d1-242a-42df-8d4d-d843fff3bffb.png)

jsnmoon replied:
Good catch, fixed.

kangzj commented on projects/plugins/jetpack/modules/search/class-jetpack-search-customberg.php:
```diff
@@ -74,13 +79,33 @@ public function jetpack_search_admin_page() {
+		if ( ! $this->wp_supports_customberg() ) {
+			?>
+				<script>
+					window.location.href="/wp-admin/customize.php?autofocus[section]=jetpack_search";
```
probably need to use `admin_url()` somehow for the sake of WP installed in a sub directory?

jsnmoon replied:
Fixed. Debugging this made me realize that the Customberg redirect hook was being used incorrectly, so that has been fixed as well.

kangzj commented on projects/plugins/jetpack/modules/search/class-jetpack-search-customberg.php:
```diff
@@ -140,6 +165,16 @@ public function load_assets_with_parameters( $path_prefix, $plugin_base_path ) {
+	 */
+	protected function wp_supports_customberg() {
+		// Must be WP 5.8 or greater.
+		return version_compare( get_bloginfo( 'version' ), '5.8', '>=' );
```
Legend @jsnmoon ! 👍

jsnmoon replied:
🙇‍♂️

kangzj commented on projects/plugins/jetpack/modules/search/customberg/components/sidebar/excluded-post-types-control.jsx:
```diff
@@ -0,0 +1,66 @@
+ *
+ * @param {object} props - component properties.
+ * @param {boolean} props.disabled - disables the control.
+ * @param {Function} props.onChange - invoked with a new color when the selected color has changed.
```
```suggestion
 * @param {Function} props.onChange - invoked with excluded post types when excluded post types have changed.
```

jsnmoon replied:
👍

kangzj commented on projects/plugins/jetpack/modules/search/customberg/components/sidebar/index.jsx:
```diff
@@ -0,0 +1,113 @@
+export default function Sidebar() {
+	const currentTab = useSelect( select => {
+		const { getActiveComplementaryArea } = select( interfaceStore );
+		return getActiveComplementaryArea( COMPLEMENTARY_AREA_SCOPE )
```
Just a thought: would we show the options as default?

jsnmoon replied:
This UX is patterned after the new Widget editor in WP 5.8, but I'd be happy to blaze our own path here.

Went ahead and changed to make the Options tab the default upon loading.

kangzj commented on projects/plugins/jetpack/modules/search/customberg/lib/analytics.js:
```diff
@@ -0,0 +1,17 @@
+export const eventPrefix = 'jetpack_search_customberg';
+
+// Force enable analytics! Ignore isAnalyticsEnabled flag used by the Instant Search application.
+export const initialize = ( ...args ) => initializeTracks( ...args, true );
```
tracking functionality could be extracted upper level, would be less coupled

jsnmoon replied:
Could you clarify what you mean here? This module is merely a small wrapper for Instant Search's analytics wrapper.

kangzj replied:
> Could you clarify what you mean here? This module is merely a small wrapper for Instant Search's analytics wrapper.

Was just thinking, analytics as a shared module would be better on the same level of instant search / customburg, rather than nested in instant search module. Definitely not a priority for now.

jsnmoon replied:
Ah, got it.

kangzj reviewed (changes requested):
It was exciting to try the new customizer out. Excellent job! 👍 
Found the following,
1. Excluded post type: it stops user from excluding more post type when there are still two options left.
2. Bonus: ask user if they want to leave the page if not saved.
3. Others are all commented.

#### EDIT: 
---------------
1. Bonus: Possibly set min height or sth?
![image](https://user-images.githubusercontent.com/1425433/127267783-a83629a8-cc0b-45d0-a1e0-2316fb6df80f.png)

2. Bonus: A link to widget editor somewhere?

3. When clicking on the cross, I got an error in console. 
![image](https://user-images.githubusercontent.com/1425433/127269894-28620abb-ad60-44f7-9085-23c8506eaf2a.png)
![image](https://user-images.githubusercontent.com/1425433/127269834-8e872c35-c069-49b8-a8b0-9de51ff3a706.png)

Commit:
Handle empty string CSVs

Commit:
Fix erroneous comment

Commit:
Use options tab as default

Commit:
Prevent modal closure in Customberg context

Commit:
Add code comment

Commit:
Add link to widget editor

Commit:
Use admin_url for redirection, fix action binding

jsnmoon commented:
> Excluded post type: it stops user from excluding more post type when there are still two options left. (@kangzj)

Fixed! This should now work as expected.

> When clicking on the cross, I got an error in console. (@kangzj)

Fixed! Interacting with the cross should result in a noop.

> Bonus: ask user if they want to leave the page if not saved. (@kangzj)

I think this is a great idea, added to code comments.

> Bonus: A link to widget editor somewhere? (@kangzj)

Added to the 'Jetpack Search' tab of the sidebar.

Commit:
Merge branch 'master' of github.com:Automattic/jetpack into add/search-customberg-2

kangzj commented on projects/plugins/jetpack/modules/search/class-jetpack-search-customberg.php:
```diff
@@ -74,13 +80,33 @@ public function jetpack_search_admin_page() {
+		if ( ! $this->wp_supports_customberg() ) {
+			?>
+				<script>
+					window.location.href="<?php echo esc_url( admin_url() ); ?>customize.php?autofocus[section]=jetpack_search";
```
```suggestion
					window.location.href="<?php echo esc_url( admin_url('customize.php?autofocus[section]=jetpack_search') ); ?>";
```

kangzj reviewed (dismissed):
Excellent work. It works and looks good to me 👍

Commit:
Use admin_url correctly

Commit:
Merge branch 'master' of github.com:Automattic/jetpack into add/search-customberg-2

Commit:
Use .min.* suffix for use in WPCOM in the future

Commit:
Omit `.bundle` from file name

Commit:
Refactor and use search sidebar printing

jeherve reviewed (changes requested):
This is looking great, nice work! ⭐ 

I tested in Chrome on a site using Twenty Twenty and spotted some potential alignment / padding issues:

![image](https://user-images.githubusercontent.com/426388/128063793-ec2b8118-c7dc-4e61-b1df-6214f0c23a73.png)

![image](https://user-images.githubusercontent.com/426388/128064145-5648ca3e-f76e-4e18-a929-80709c16d4e6.png)

![image](https://user-images.githubusercontent.com/426388/128064320-6bb1b479-1d10-4d9b-830a-a295f7c70df6.png)

Should there be more space around those elements?

![image](https://user-images.githubusercontent.com/426388/128063944-baea00ae-e173-41bf-9417-e77cbb571813.png)

Should we have a Jetpack logo here?
![image](https://user-images.githubusercontent.com/426388/128064197-a54685a1-762e-44e0-a10e-fa021347d93f.png)

bluefuton committed:
Add jetpack domain to translations

Co-authored-by: Jeremy Herve <jeremy@jeremy.hu>

bluefuton commented:
Added the new babel-preset-default package separately in #20634.

bluefuton commented on projects/plugins/jetpack/package.json:
```diff
@@ -23,7 +23,7 @@
 		"build-php": "composer install --ignore-platform-reqs",
 		"build-production": "pnpm run distclean && pnpm run install-if-deps-outdated && pnpm run build-production-client && pnpm run build-production-php && pnpm run build-production-extensions && pnpm run build-production-search && pnpm run build-asset-cdn-json",
-		"build-production-concurrently": "pnpm run distclean && pnpm run install-if-deps-outdated && pnpx concurrently 'pnpm:build-production-client' 'pnpm:build-production-php' 'NODE_ENV=production pnpm run build-extensions' 'NODE_ENV=production pnpm run build-search' && pnpm run build-asset-cdn-json",
+		"build-production-concurrently": "pnpm run distclean && pnpm run install-if-deps-outdated && pnpx concurrently 'pnpm:build-production-client' 'pnpm:build-production-php' 'NODE_ENV=production pnpm run build-extensions' && NODE_ENV=production pnpm run build-search && pnpm run build-asset-cdn-json",
```
@jsnmoon has advised that this change is no longer required.

bluefuton commented on projects/plugins/jetpack/package.json:
```diff
@@ -165,6 +165,7 @@
 		"@testing-library/react": "11.2.7",
 		"@testing-library/react-hooks": "4.0.1",
 		"@testing-library/user-event": "12.8.3",
+		"@wordpress/babel-preset-default": "6.2.0",
```
This has been added in a separate PR.

jsnmoon replied:
Specifically, https://github.com/Automattic/jetpack/pull/20634.

Commit:
Revert package.json changes

Commit:
Merge branch 'master' of github.com:Automattic/jetpack into add/search-customberg-2

Commit:
Use global to fetch WP version

Commit:
Make comment easier to understand

Commit:
Tweak control spacing

Commit:
Use Jetpack colophon in sidebar description

Commit:
Use correct import for portal creation

Commit:
Merge branch 'master' of github.com:Automattic/jetpack into add/search-customberg-2

jsnmoon commented:
> Should there be more space around those elements? (@jeherve)
> 
> ![image](https://user-images.githubusercontent.com/426388/128063944-baea00ae-e173-41bf-9417-e77cbb571813.png)

Fixed! Increased spacing between inputs where appropriate.

> Should we have a Jetpack logo here? (@jeherve)
> ![image](https://user-images.githubusercontent.com/426388/128064197-a54685a1-762e-44e0-a10e-fa021347d93f.png)

I hadn't thought of this before, but yes! Fixed; now uses the green Jetpack colophon.

Commit:
Bump ES check from 9 to 10

Commit:
Hide wp-admin notices

jsnmoon commented:
> I tested in Chrome on a site using Twenty Twenty and spotted some potential alignment / padding issues: (@jeherve)
> 
> ![image](https://user-images.githubusercontent.com/426388/128063793-ec2b8118-c7dc-4e61-b1df-6214f0c23a73.png)
> 
> ![image](https://user-images.githubusercontent.com/426388/128064145-5648ca3e-f76e-4e18-a929-80709c16d4e6.png)
> 
> ![image](https://user-images.githubusercontent.com/426388/128064320-6bb1b479-1d10-4d9b-830a-a295f7c70df6.png)

I've tried reproducing these issues locally and in Jurassic Ninja without much success, unfortunately. Could you provide additional repro steps? And could we possibly punt this to a future PR?

jeherve committed:
Merge remote-tracking branch 'origin/master' into add/search-customberg-2

jeherve committed:
Remove unnecessary import

jeherve reviewed (dismissed):
> I've tried reproducing these issues locally and in Jurassic Ninja without much success, unfortunately. Could you provide additional repro steps? And could we possibly punt this to a future PR?

It seems to happen all the time in Chrome for me, as soon as I open the settings screen. I'm happy to punt this to a follow-up PR though, indeed.

Approving this for now.

jsnmoon commented on projects/plugins/jetpack/modules/search/customberg/lib/_base-styles.scss:
```diff
@@ -0,0 +1,29 @@
+$gray-800: #2f2f2f;
+$gray-300: #ddd;
+
+@import '~@wordpress/base-styles/variables';
+@import '~@wordpress/base-styles/breakpoints';
+@import '~@wordpress/base-styles/mixins';
+@import '~@wordpress/base-styles/animations';
```
That makes a lot of sense 👍

jsnmoon commented:
Deployed in r230498-wpcom. 

Some adjustments were necessary for the deployment; they will be backported to Jetpack in a future PR.
